File: Backtracking/CS.py
class Solution:
    def combinationSum(self, candidates, target):
        ans = []
        temp = []
        self.backtrack(candidates, target, ans, temp, 0)
        return ans

    def backtrack(self, candidates, target, ans, temp, index):
        if target < 0:
            return
        elif target == 0:
            ans.append(list(temp))
        else:
            for i in range(index, len(candidates)):
                temp.append(candidates[i])
                self.backtrack(candidates, target - candidates[i], ans, temp, i)
                temp.pop()

CS = Solution()
assert CS.combinationSum([2,3,6,7], 7) == [[2,2,3], [7]], "ex1 failed"
assert CS.combinationSum([2,3,5], 8) == [[2,2,2,2],[2,3,3],[3,5]], "ex2 failed"
assert CS.combinationSum([2], 1) == [], "ex3 failed"
print("all pass")

File: Backtracking/39.Combination Sum/CS.cpp
/*
backtracking
candidates = [2,3,6,7] target = 7
            2
        2      3 ...
    2; 3 = 7 bingo append to answer, 6, 7 return           
2 = 8 return, 3, 6, 7, return  
*/
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> ans;
        vector<int> temp;
        backtrack(candidates, target, temp, ans, 0);
        return ans;
    }
    void backtrack(vector<int>& cand, int target, vector<int>& temp, vector<vector<int>> &ans, int index) {
        if(target < 0)
            return;
        else if(target == 0) {
            ans.push_back(temp); // ans = [2,2,3]
        } else {
            for(int i = index; i < cand.size(); i++) { // i = 0; 0; ...; 0, 1
                temp.push_back(cand[i]);               
                // temp [2]; [2,2]; ...; [2,2,2,2]; [2,2,2,3]; ... [2,2,3]
                backtrack(cand, target - cand[i], temp, ans, i);
                temp.pop_back();                
                // temp [2,2,2]; [2,2,2]; ... [2,2]
            }
        }
    }
};